pkg/liveshare/rpc.go

```diff
@@ -40,7 +42,60 @@ func (r *rpcClient) do(ctx context.Context, method string, args, result interfac
 	return waiter.Wait(waitCtx, result)
 }
 
-type nullHandler struct{}
+type handlerFn func(conn *jsonrpc2.Conn, req *jsonrpc2.Request)
 
-func (nullHandler) Handle(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) {
+type requestHandler struct {
+	handlersMu sync.RWMutex
+	handlers   map[string][]handlerFn
+}
+
+func newRequestHandler() *requestHandler {
+	return &requestHandler{handlers: make(map[string][]handlerFn)}
+}
+
+func (r *requestHandler) register(requestType string, handler handlerFn) func() {
+	r.handlersMu.Lock()
+	defer r.handlersMu.Unlock()
+
+	if _, ok := r.handlers[requestType]; !ok {
+		r.handlers[requestType] = []handlerFn{}
+	}
+
+	r.handlers[requestType] = append(r.handlers[requestType], handler)
+
+	return func() {
+		r.deregister(requestType, handler)
+	}
+}
+
+func (r *requestHandler) deregister(requestType string, handler handlerFn) {
```

Comment:
Functions are not comparable in Go, so there's no way to ask "is handler equal to some other value?"

Instead you'll need to identify handlers by some other means, such as a pointer to a struct:
```
// types
type struct handler { fn func(conn *jsonrpc2.Conn, req *jsonrpc2.Request) }

type rpcClient struct {
   ...
   handlersMu sync.Mutex
   handlers   map[string][]*handler
}

func (cli *rpcClient) register(t, fn) func() {
  h := &handler{fn}
  cli.handlers[t] = append(cli.handlers[t], h)
  return func() { cli.deregister(t, h) }
}

func (cli *rpcClient) deregister(h *handler) {
   ...remove h from cli.handlers[t]...
}
```

Reply:
changed in latest commit